### PR TITLE
feat: Implemented new Db APIs for SQLite

### PR DESCRIPTION
**Pull request by nicolatimeus**
Signed-off-by: Nicola Timeus <nicola.timeus@eurotech.com>

> **Note**: We are using the Conventional Commits convention for our pull request titles. Please take a look at the [PR title format document](https://github.com/eclipse/kura/blob/develop/CONTRIBUTING.md#submitting-the-changes) for the supported [types](https://github.com/eclipse/kura/blob/develop/CONTRIBUTING.md#type) and [scopes](https://github.com/eclipse/kura/blob/develop/CONTRIBUTING.md#scope).

Brief description of the PR. [e.g. Added `null` check on `object` to avoid `NullPointerException`]

* Implemented new Db APIs for SQLite
* Fixed possible issue with H2 message store identity generator reset
* H2 Wire Record store implementation now always drops and recreates columns if inserted property type is not compatible

### CODE REVIEW

File: kura/org.eclipse.kura.db.sqlite.provider/src/main/java/org/eclipse/kura/internal/db/sqlite/provider/SqliteMessageStoreImpl.java
Comment (marcellorinaldo) [inline kura/org.eclipse.kura.db.sqlite.provider/src/main/java/org/eclipse/kura/internal/db/sqlite/provider/SqliteMessageStoreImpl.java]: Does it make sense to have a "`withPreparedStatement`" instead of `withConnection` so to factor in some more common code?

Comment (marcellorinaldo) [reply kura/org.eclipse.kura.db.sqlite.provider/src/main/java/org/eclipse/kura/internal/db/sqlite/provider/SqliteMessageStoreImpl.java]: Something like:
```java
private interface PreparedStatementCallable<T> {
    public T call(Connection conn, PreparedStatement preparedStatement) throws SQLException;
}

private <T> T withPreparedStatement(String sql, String exceptionMessage,
            final PreparedStatementCallable<T> callable) throws KuraStoreException {
    try (final Connection connection = this.dbService.getConnection()) {
        try (PreparedStatement preparedStatement = connection.prepareStatement(sql)) {
            return callable.call(connection, preparedStatement);
        }
    } catch (final Exception e) {
        throw new KuraStoreException(e, exceptionMessage);
    }
}
```

File: kura/org.eclipse.kura.db.sqlite.provider/src/main/java/org/eclipse/kura/internal/db/sqlite/provider/SqliteMessageStoreImpl.java
Comment (marcellorinaldo) [inline kura/org.eclipse.kura.db.sqlite.provider/src/main/java/org/eclipse/kura/internal/db/sqlite/provider/SqliteMessageStoreImpl.java]: Should we return 0 here?

Comment (nicolatimeus) [reply kura/org.eclipse.kura.db.sqlite.provider/src/main/java/org/eclipse/kura/internal/db/sqlite/provider/SqliteMessageStoreImpl.java]: I agree that -1 is not ok here, I have moved the code to a utility class and have it throw an exception instead since it seems that H2 and SQLite always return a result set for this query.

File: kura/org.eclipse.kura.db.sqlite.provider/src/main/java/org/eclipse/kura/internal/db/sqlite/provider/SqliteMessageStoreImpl.java
Comment (nicolatimeus) [reply kura/org.eclipse.kura.db.sqlite.provider/src/main/java/org/eclipse/kura/internal/db/sqlite/provider/SqliteMessageStoreImpl.java]: I have added this to a utility class

File: kura/org.eclipse.kura.util/src/main/java/org/eclipse/kura/util/message/store/SqlMessageStoreHelper.java
Comment (marcellorinaldo) [inline kura/org.eclipse.kura.util/src/main/java/org/eclipse/kura/util/message/store/SqlMessageStoreHelper.java]: I think we need to set the value to null if the condition is false.

File: kura/org.eclipse.kura.util/src/main/java/org/eclipse/kura/util/message/store/SqlMessageStoreHelper.java
Comment (marcellorinaldo) [inline kura/org.eclipse.kura.util/src/main/java/org/eclipse/kura/util/message/store/SqlMessageStoreHelper.java]: As for in one of the previous comments, should we throw an exception instead of returning -1?

File: kura/org.eclipse.kura.db.sqlite.provider/src/main/java/org/eclipse/kura/internal/db/sqlite/provider/SqliteWireRecordStoreImpl.java
Comment (marcellorinaldo) [inline kura/org.eclipse.kura.db.sqlite.provider/src/main/java/org/eclipse/kura/internal/db/sqlite/provider/SqliteWireRecordStoreImpl.java:49]: Is there a particular reason for using `BIGINT` instead of `DATETIME` as the timestamp type?

Comment (nicolatimeus) [reply kura/org.eclipse.kura.db.sqlite.provider/src/main/java/org/eclipse/kura/internal/db/sqlite/provider/SqliteWireRecordStoreImpl.java:49]: I have kept it in this way for consistency with the old DbStore and H2DbWireRecordStore implementations. I don't remember the reason for having the timestamp as an integer.

**Review by marcellorinaldo — DISMISSED** [review]

File: kura/org.eclipse.kura.db.sqlite.provider/src/main/java/org/eclipse/kura/internal/db/sqlite/provider/SqliteMessageStoreImpl.java
Comment (marcellorinaldo) [inline kura/org.eclipse.kura.db.sqlite.provider/src/main/java/org/eclipse/kura/internal/db/sqlite/provider/SqliteMessageStoreImpl.java]: A `createTable` without index creation could be helpful in the helper

File: kura/org.eclipse.kura.util/src/main/java/org/eclipse/kura/util/message/store/SqlMessageStoreQueries.java
Comment (marcellorinaldo) [inline kura/org.eclipse.kura.util/src/main/java/org/eclipse/kura/util/message/store/SqlMessageStoreQueries.java:35]: These required "index" fields need to be set also if someone don't wants to invoke the "createIndexes" method. Is that on purpose?

Comment (nicolatimeus) [reply kura/org.eclipse.kura.util/src/main/java/org/eclipse/kura/util/message/store/SqlMessageStoreQueries.java:35]: Currently all fields are checked to ensure that the user explicitly reviews them at least once and provide a value for each of them. If a field is not used, any value can be set to it.

File: kura/org.eclipse.kura.util/src/main/java/org/eclipse/kura/util/message/store/SqlMessageStoreHelper.java
Comment (marcellorinaldo) [inline kura/org.eclipse.kura.util/src/main/java/org/eclipse/kura/util/message/store/SqlMessageStoreHelper.java]: A subtle thing can happen here: the long will be set as long in the underlying `setObject()` of the execute method. Some databases might not automatically reparse this as a `Timestamp`. So I would replace this line with:

```java
final long difference = System.currentTimeMillis() - TimeUnit.SECONDS.toMillis(purgeAge);
Timestamp timestamp = new Timestamp(difference);
```

**Comment by MMaiero** [comment]
@marcellorinaldo @salvatore-coppola Please provide your review on this PR.
Once we have review approvals we will merge the code.
The issue reported by Sonar is on code duplication that unfortunately cannot be solved right now due to the fact that the code for SQLite is indeed similar to H2. But H2 is going to be removed in future, so the duplication is acceptable for now.

**Comment by marcellorinaldo** [comment]
> @marcellorinaldo @salvatore-coppola Please provide your review on this PR.
> Once we have review approvals we will merge the code.
> The issue reported by Sonar is on code duplication that unfortunately cannot be solved right now due to the fact that the code for SQLite is indeed similar to H2. But H2 is going to be removed in future, so the duplication is acceptable for now.

Finishing to review the wire part. It will be due soon.

**Review by marcellorinaldo — APPROVED** [review]

**Pull request merged**
